### PR TITLE
Change URI to Builder pattern

### PR DESCRIPTION
**Pull request by adienzel**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,5 @@ cmake_install.cmake
 /.ninja_deps
 /.ninja_log
 .vscode
+/uprotocol-core-api/
```
Comment (xz62px) [inline .gitignore:24]: revert

File: CMakeLists.txt [hunk 2]
```diff
@@ -20,9 +20,13 @@
 
 cmake_minimum_required(VERSION 3.18.0)
 file(STRINGS "VERSION" VERSION_STRING)
+
+ set(CMAKE_CXX_STANDARD 17)
 project(uprotocol-cpp VERSION "${VERSION_STRING}" LANGUAGES CXX)
 
-set(VCPKG $ENV{HOME}/vcpkg/scripts/buildsystems/vcpkg.cmake)
+set(CMAKE_CXX_FLAGS "${CAMKE_CXX_FLAGS} -g -L/usr/local/lib -I../ -pthread -Wall -Wpedantic")
```
Comment (xz62px) [inline CMakeLists.txt:27]: revert

File: include/uprotocol-cpp/uri/builder/BuildEntity.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#include <algorithm>
+#include <cctype>
+#include <optional>
+#include <string>
+#include <array>
+#include <string_view>
+#include <arpa/inet.h>
+#include <spdlog/spdlog.h>
+#include "../tools/Utils.h"
+#include "../uprotocol-core-api/src/main/proto/uri.pb.h"
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/builder/BuildEntity.h:41]: should be #include <src/main/proto/uri.pb.h>

File: include/uprotocol-cpp/uri/builder/BuildEntity.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+#include <algorithm>
+#include <cctype>
+#include <optional>
+#include <string>
+#include <array>
+#include <string_view>
+#include <arpa/inet.h>
+#include <spdlog/spdlog.h>
+#include "../tools/Utils.h"
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/builder/BuildEntity.h:40]: should be #include<uprotocol-cpp/uri/tools/>

File: include/uprotocol-cpp/uri/builder/BuildUAuthority.h [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+#include <algorithm>
+#include <cctype>
+#include <optional>
+#include <string>
+#include <array>
+#include <string_view>
+#include <arpa/inet.h>
+#include <spdlog/spdlog.h>
+#include "../tools/Utils.h"
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/builder/BuildUAuthority.h:40]: same comments

File: include/uprotocol-cpp/uri/builder/BuildUResource.h [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+#include <algorithm>
+#include <cctype>
+#include <optional>
+#include <string>
+#include <array>
+#include <string_view>
+#include <arpa/inet.h>
+#include <spdlog/spdlog.h>
+#include "../tools/Utils.h"
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/builder/BuildUResource.h:40]: same comments

File: include/uprotocol-cpp/uri/datamodel/UUri.h [hunk 7]
```diff
@@ -22,15 +22,16 @@
...
+#ifndef UURI_H_
+#define UURI_H_
 
 #include <string>
 #include <utility>
 #include "UriFormat.h"
 #include "UAuthority.h"
 #include "UEntity.h"
 #include "UResource.h"
+#include "../uprotocol-core-api/src/main/proto/uri.pb.h"
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/datamodel/UUri.h:34]: same comment

File: include/uprotocol-cpp/uri/datamodel/UUri.h [hunk 8]
```diff
@@ -158,14 +172,36 @@ class UUri : public UriFormat {
...
+        if (!u_authority.isEmpty()) {
+            uUri_.mutable_authority()->CopyFrom(u_authority.getProtobufAuthority());
+        }
+        if (!u_entity.isEmpty()) {
+            uUri_.mutable_entity()->CopyFrom(u_entity.getProtobufEntity());
+        }
+        if (!u_resource.isEmpty()) {
+            uUri_.mutable_resource()->CopyFrom(u_resource.getProtoUresource());
+        }
+        //todo hash
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/datamodel/UUri.h:197]: Dont forget the dodo

File: include/uprotocol-cpp/uri/serializer/IpAddress.h [hunk 9]
```diff
@@ -53,33 +55,37 @@ class IpAddress {
     /**
      * Constructor with IP address in byte format.
      */
-    IpAddress(std::vector<uint8_t> const& ipBytes, AddressType type) : type_(type), ipBytes_(ipBytes)  {
+    IpAddress(std::vector<uint8_t> const& ipBytes, AddressType type) : type_(type) , ipBytes_(ipBytes) {
+//        for (unsigned char ipByte : ipBytes) {
+//            ipBytes_.push_back(ipByte);
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/serializer/IpAddress.h:60]: delete if not needed

File: include/uprotocol-cpp/uri/serializer/IpAddress.h [hunk 10]
```diff
@@ -53,33 +55,37 @@ class IpAddress {
...
      * Get the type of IP address.
      */
-    AddressType getType() const { return type_; }
+    auto getType() const { return type_; }
 
     /**
      * Get the string format of IP address.
      */
-    std::string getString() const { return ipString_; }
+    auto getString() const { return ipString_; }
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/serializer/IpAddress.h:74]: we know the type , why use auto and not std::string?

File: include/uprotocol-cpp/uri/serializer/IpAddress.h [hunk 11]
```diff
@@ -53,33 +55,37 @@ class IpAddress {
...
      * Get the byte format of IP address.
      */
-    std::vector<uint8_t> getBytes() const { return ipBytes_; }
+    auto getBytes() const { return ipBytes_; }  
 
     /**
      * Number of bytes in IPv4 address.
      */
-    static constexpr uint8_t IPV4_ADDRESS_BYTES = 4;
+    static constexpr uint8_t IpV4AddressBytes = 4;
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/serializer/IpAddress.h:84]: add _

File: include/uprotocol-cpp/uri/serializer/LongUriSerializer.h [hunk 12]
```diff
@@ -44,14 +46,14 @@ class LongUriSerializer {
      * @return Returns the String format of the supplied UUri that can be used as a sink or a source
      * in a uProtocol publish communication.
      */
-    static std::string serialize(const UUri& uUri);
+    static auto serialize(const v1::UUri& u_uri) -> std::string;
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/serializer/LongUriSerializer.h:49]: same comment

File: include/uprotocol-cpp/uri/serializer/LongUriSerializer.h [hunk 13]
```diff
@@ -22,13 +22,15 @@
...
 
-#include <uprotocol-cpp/uri/datamodel/UUri.h>
-#include <uprotocol-cpp/uri/datamodel/UAuthority.h>
-#include <uprotocol-cpp/uri/datamodel/UEntity.h>
-#include <uprotocol-cpp/uri/datamodel/UResource.h>
+#include <uprotocol-cpp/uri/builder/BuildUResource.h>
+#include <uprotocol-cpp/uri/builder/BuildEntity.h>
+#include <uprotocol-cpp/uri/builder/BuildUAuthority.h>
+#include <uprotocol-cpp/uri/builder/BuildUUri.h>
+#include "../uprotocol-core-api/src/main/proto/uri.pb.h"
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/serializer/LongUriSerializer.h:32]: please fix the includes

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h [hunk 14]
```diff
@@ -40,44 +46,79 @@ class MicroUriSerializer {
...
+        
+        /**
+         * The length of a local micro URI.
+         */
+    static constexpr uint32_t LocalMicroUriLength = 8;
     /**
      * The length of a IPv4 micro URI.
      */
-    static constexpr uint32_t IPV4_MICRO_URI_LENGTH = 12;
+    static constexpr uint32_t IpV4MicroUriLength = 12;
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:73]: add _ to all

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h [hunk 15]
```diff
@@ -40,44 +46,79 @@ class MicroUriSerializer {
...
+    
+    [[nodiscard]] [[maybe_unused]] auto isMicroForm(const uprotocol::v1::UEntity &entity) -> bool {
+        return entity.has_id() && entity.id() > 0;
+    }
+    
+    
+    [[nodiscard]] [[maybe_unused]] auto isMicroForm(const uprotocol::v1::UAuthority &authority) -> bool {
+        return isEmpty(authority) || (authority.has_ip() && !authority.ip().empty())  || (authority.has_id() && !authority.id().empty());
+    }
+
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:114]: remove one space

File: include/uprotocol-cpp/uri/tools/Utils.h [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+#include <algorithm>
+#include <cctype>
+#include <optional>
+#include <string>
+#include <array>
+#include <string_view>
+#include <arpa/inet.h>
+#include <spdlog/spdlog.h>
+
```
Comment (xz62px) [inline include/uprotocol-cpp/uri/tools/Utils.h:40]: remove empty line and fix include

File: src/uri/LongUriSerializer.cpp [hunk 17]
```diff
@@ -23,40 +23,39 @@
...
-#include <uprotocol-cpp/uri/datamodel/UResource.h>
+#include <iostream>
+#include <uprotocol-cpp/uri/builder/BuildUUri.h>
+#include <uprotocol-cpp/uri/builder/BuildUAuthority.h>
+#include <uprotocol-cpp/uri/builder/BuildEntity.h>
+#include <uprotocol-cpp/uri/builder/BuildUResource.h>
 #include <uprotocol-cpp/uri/serializer/LongUriSerializer.h>
 
-using namespace uprotocol::uri;
+//using namespace uprotocol::uri;
```
Comment (xz62px) [inline src/uri/LongUriSerializer.cpp:36]: delete if not needed

File: test/uri/builder/UAuthorityTest.cpp [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ 
+ 
+ */
+#include <string>
```
Comment (xz62px) [inline test/uri/builder/UAuthorityTest.cpp:28]: please modify the test to gtest instead of cgreen

File: test/uri/builder/UEntityTest.cpp [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ 
+ 
+ */
+#include <string>
```
Comment (xz62px) [inline test/uri/builder/UEntityTest.cpp:28]: please modify the test to gtest instead of cgreen

File: test/uri/builder/UUriTest.cpp [hunk 20]
```diff
@@ -0,0 +1,382 @@
...
+ */
+#include <string>
+#include <cgreen/cgreen.h>
+
+#include "uprotocol-cpp/uri/builder/BuildUUri.h"
+#include "uprotocol-cpp/uri/builder/BuildEntity.h"
+#include "uprotocol-cpp/uri/builder/BuildUResource.h"
+#include "uprotocol-cpp/uri/builder/BuildUAuthority.h"
+#include "uprotocol-cpp/uri/serializer/LongUriSerializer.h"
+#include "uprotocol-cpp/uri/serializer/MicroUriSerializer.h"
```
Comment (xz62px) [inline test/uri/builder/UUriTest.cpp:36]: please modify the test to gtest instead of cgreen

**Review by xz62px — CHANGES_REQUESTED** [review]

**Comment by stevenhartley** [comment]
Closing PR, will use @MelamudMichael branch instead

**Pull request closed without merging**